lib/req/steps.ex

```diff
@@ -1932,13 +1932,19 @@ defmodule Req.Steps do
         redirect_count = Req.Request.get_private(request, :req_redirect_count, 0)
 
         if redirect_count < max_redirects do
-          request =
-            request
-            |> build_redirect_request(response)
-            |> Req.Request.put_private(:req_redirect_count, redirect_count + 1)
-
-          {_, result} = Req.Request.run(request)
-          {Req.Request.halt(request), result}
+          case Req.Response.get_header(response, "location") do
+            [location | _] ->
+              request =
+                request
+                |> build_redirect_request(response, location)
+                |> Req.Request.put_private(:req_redirect_count, redirect_count + 1)
+
+              {_, result} = Req.Request.run(request)
+              {Req.Request.halt(request), result}
+
+            _ ->
+              {Req.Request.halt(request), response}
```

Comment:
we only want to halt when we make a new request (Req.Request.run) or raise. At the beginning of the step we have a `redirect?` conditional, basically whether this step is a no-op. Could you make the change there instead? Perhaps something like:

```
with true <- response.status in ...,
     [location | _] <- ... do
  ...
else
  _ ->
    {request, response}
end
```

Reply:
Ok great, that's much simpler.  Also, with that in mind it looks like we could refactor up another step (to the `cond`).  I just updated branch with suggested change and also added a test for when :redirect is false.